### PR TITLE
Move generic HashTreeRoot funcs to //shared

### LINKED ISSUE #5222: Refactor Stateutils to a Shared Library
Discussed in #5219 with @terencechain , currently stateutils has a lot of HashTreeRoot methods for non-state objects. These methods would be better lived in a shared library rather than in `stateutils` which is mostly utilities for hashing the beacon state trie. This issue is to track the refactor and clean up of this.

### PR DESCRIPTION
**Pull request by Pablosan**
**What type of PR is this?**

Other (enhancement)

**What does this PR do? Why is it needed?**

There are several methods in `stateutils` that would be better suited to a shared library, as they are not directly related to HashTreeRoot states. This PR moves these generic methods to `//shared/htrutils` along with a couple more dependent files.

**Which issues(s) does this PR fix?**

Addresses #5222

**Other notes for review**
  - [X] Move generic methods into a separate `htrutils.go` file, in preparation for moving
  - [x] Move `htrutils.go`, `hash_function.go`, `helpers.go`, and `merkleize.go` from `stateutils` to `//shared/htrutils`
  - [x] Update references to methods in `htrutils.go`, `hash_function.go`, `helpers.go`, and `merkleize.go` to reference their new home
  - [x] Add tests for `htrutils.go`, `hash_function.go`, `helpers.go` and `merkleize.go`

### CODE REVIEW

**Comment by Pablosan** [comment]
@nisdas I still need to add tests and run the validator locally before this is ready to be merged, but if you can find the time I wouldn't mind getting your first impression from a quick review of what I have so far. Thanks in advance!

**Comment by nisdas** [comment]
@Pablosan On a first look, this all looks good to me :+1: , it seems to be on the right track.

**Comment by Pablosan** [comment]
Thanks for the feedback @nisdas! Actually, I think this work might go on for a while so I have a suggestion: can we just review what I have so far in an effort to merge it sooner rather than later? I have removed "WIP" from the title and also changed "Fixes #5222" to "Addresses #5222" so that the Issue doesn't get auto-closed.

That way we can avoid any potentially nasty master merges (either for me or someone else), and I can continue the work in another PR.

Thoughts?

File: beacon-chain/state/state_trie.go [hunk 1]
```diff
@@ -203,7 +204,7 @@ func (b *BeaconState) HashTreeRoot(ctx context.Context) ([32]byte, error) {
 // pads the leaves to a power-of-two length.
 func merkleize(leaves [][]byte) [][][]byte {
 	hashFunc := hashutil.CustomSHA256Hasher()
-	layers := make([][][]byte, stateutil.GetDepth(uint64(len(leaves)))+1)
+	layers := make([][][]byte, htrutils.GetDepth(uint64(len(leaves)))+1)
```
Comment (0xKiwi) [inline beacon-chain/state/state_trie.go:207]: When do we decide to name `htrutils` vs. `htrutil`? Singular vs. Plural?

Comment (Pablosan) [reply beacon-chain/state/state_trie.go:207]: @0xKiwi I didn't see a clear pattern in `//shared` so I just picked one. Some of the shared packages are singular, some are plural. Is there a reason to choose one over the other? I'm more than happy to make it `htrutil`. I have no preference.

Comment (nisdas) [reply beacon-chain/state/state_trie.go:207]: I think `htrutils` is fine here, the plural form makes sense for this.

Comment (Pablosan) [reply beacon-chain/state/state_trie.go:207]: `//shared/grpcutils` makes it even more confusing. Not only is there only on file, the file only has one function, so ¯\_(ツ)_/¯.

**Comment by Pablosan** [comment]
@nisdas I have added tests for all four files moved to `//shared/htrutils` so I think this PR is ready to go. I added another To Do item to revisit `stateutil` and determine what else, if anything, should be moved. I also changed "Fixes" to "Addresses" in reference to the associated issue, which should leave the issue opened so I can continue to reference new PR's to it.

Ready for review!

**Pull request merged**